# Importing Libraries

In [7]:
import pandas as pd
import numpy as np
import os

# Importing Data

In [8]:
# Creating a path to project folder.

path = r'C:\Users\TanaT\(CF) Achievement 4 - Instacart Basket Analysis'

In [10]:
# Importing exported data from previous exercise (4.7).

df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_derived_variables.pkl'))

In [11]:
# Checking shape of dataframe. 

df_ords_prods_merge.shape

(32404859, 18)

In [12]:
# Checking dataframe.

df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     NaN                 10          0   both   
4                     9.0                 11          1   both   

     price_range_loc   busiest_day    busiest_days busiest_period_of_day  
0  Mid-range product  Regular busy  Regularly busy           Most orders  
1  Mid-range product  Regular busy  Regularly busy        Average orders  
2  Mid-range product   Busiest day    Busiest days        Average orders  
3  Mid-range product  Regular busy    Slowest days           Most orders  
4  Mid-range product    Least busy    Slowest days        Average orders

# Task

### Step 2 - Find the aggregated mean of the 'order_number' column grouped by 'department_id' for the entire dataframe.

In [13]:
# Grouping data by 'department_id'.
# Applying .agg() function to the 'order_number' column to find the mean number of orders placed.
# Finding the average number of orders per user for each department ID. 

df_ords_prods_merge.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### Step 3 - Analyze the result. How do the results for the entire dataframe differ from those of the subset?

This code gave us the average number of orders per user for each department id, to know which department(s) sell better.

When comparing the means for the department ids of the first million rows (subset) vs. the entire dataframe, the means appear to be similar.

Except for a few department ids, the means do not appear much different, but a t-test could be run to see if the means are significantly different.

For both the entire dataframe and the subset, the mean of department 21 was the highest (25.53 for subset).

The department ids with the biggest differences is department 10 and 21. The mean of department 10 for the entire dataframe is higher by about 1.5 points compared to the subset mean (18.68). While the mean of department 21 in the entire dataframe has over a 2 point difference compared to the subset mean (25.53) and is lower.

I think overall, it's surprising that the small subset is almost representative of the entire dataframe. There are differences between the two sets of means, but it actually gave a good sense of where the entire data is (at least when calculating the means for this particular column).

### Step 4 - Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform and loc functions.¶

In [14]:
# Creating a new column ('max_order').
# Grouping user IDs and selecting column 'order_number' to transform to find the max order number for each user (for every row).
# Putting the results of the transform function into the new column.

df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\TanaT\AppData\Local\Temp\ipykernel_78304\1995621429.py:5: FutureWarning: The provided callable <function max at 0x00000167C75CB4C0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [15]:
# Fixing code.

df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform('max')

In [16]:
# Checking dataframe.

df_ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   

    days_since_prior_order  add_to_cart_order  reordered _merge  \
0                      3.0                  5          0   both   
1                     20.0                  1          1   both   
2                      6.0                 20          0   both   
3                      NaN                 10          0   both   
4                      9.0                 11          1   both   
5                     26.0                  7          0   both   
6                     30.0                  2          0   both   
7                     30.0                  1          0   both   
8                     19.0                  1          0   both   
9                     15.0                  2          1   both   
10                     NaN                  1          0   both   
11                     7.0                  1          0   both   
12                     7.0                  3          1   both   
13                    12.0                  4          1   both   
14                     2.0                  8          0   both   

      price_range_loc   busiest_day    busiest_days busiest_period_of_day  \
0   Mid-range product  Regular busy  Regularly busy           Most orders   
1   Mid-range product  Regular busy  Regularly busy        Average orders   
2   Mid-range product   Busiest day    Busiest days        Average orders   
3   Mid-range product  Regular busy    Slowest days           Most orders   
4   Mid-range product    Least b

In [17]:
# Making a loyalty flag/label based on the 'max_order' column created.

df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [18]:
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [19]:
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [20]:
# Checking count of flags.

df_ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

Note: This shows that most of the customers fall into the 'Regular customer' category.

In [21]:
# Checking dataframe with the new column.

df_ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   

    days_since_prior_order  add_to_cart_order  reordered _merge  \
0                      3.0                  5          0   both   
1                     20.0                  1          1   both   
2                      6.0                 20          0   both   
3                      NaN                 10          0   both   
4                      9.0                 11          1   both   
5                     26.0                  7          0   both   
6                     30.0                  2          0   both   
7                     30.0                  1          0   both   
8                     19.0                  1          0   both   
9                     15.0                  2          1   both   
10                     NaN                  1          0   both   
11                     7.0                  1          0   both   
12                     7.0                  3          1   both   
13                    12.0                  4          1   both   
14                     2.0                  8          0   both   

      price_range_loc   busiest_day    busiest_days busiest_period_of_day  \
0   Mid-range product  Regular busy  Regularly busy           Most orders   
1   Mid-range product  Regular busy  Regularly busy        Average orders   
2   Mid-range product   Busiest day    Busiest days        Average orders   
3   Mid-range product  Regular busy    Slowest days           Most orders   
4   Mid-range product    Least b

In [22]:
# Checking only columns about the user id, loyalty flags, and order number.

df_ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0       138  Regular customer            28
1       138  Regular customer            30
2       709      New customer             2
3       764      New customer             1
4       764      New customer             3
5       777  Regular customer            16
6       825      New customer             3
7       910  Regular customer            12
8      1052  Regular customer            10
9      1052  Regular customer            15
10     1379      New customer             1
11     1480    Loyal customer            36
12     1480    Loyal customer            38
13     1480    Loyal customer            39
14     1494  Regular customer             6
15     1494  Regular customer             7
16     1494  Regular customer             9
17     1540  Regular customer             4
18     1540  Regular customer             5
19     1540  Regular customer             6
20     1540  Regular customer             7
21     1540  Regular customer             8
22     1540  Regular customer             9
23     1540  Regular customer            14
24     1540  Regular customer            15
25     1540  Regular customer            18
26     1540  Regular customer            19
27     1540  Regular customer            20
28     1540  Regular customer            21
29     1540  Regular customer            22
30     1540  Regular customer            24
31     1540  Regular customer            25
32     1540  Regular customer            27
33     1540  Regular customer            28
34     1598      New customer             1
35     1598      New customer             2
36     1598      New customer             3
37     1647      New customer             1
38     1647      New customer             4
39     1647      New customer             6
40     1647      New customer             8
41     1647      New customer             9
42     2833      New customer             1
43     2850  Regular customer             7
44     2857  Regular customer            24
45     3010  Regular customer             5
46     3029      New customer             1
47     3395      New customer             4
48     3904      New customer             1
49     4006    Loyal customer             2
50     4006    Loyal customer             3
51     4006    Loyal customer             5
52     4006    Loyal customer             9
53     4006    Loyal customer            12
54     4122  Regular customer            11
55     4141      New customer             2
56     4343      New customer             1
57     4553      New customer             3
58     4845  Regular customer             6
59     6866      New customer             1

### Step 5 - Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category.

In [23]:
# Trying to determine whether the prices of the products purchased by loyal customers differ from those purchased by regular or new customers. 

df_ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

Note:

- There could be a problem with the max price (outlier perhaps).
- All of the groups share a minimum price of 1 dollar.
- It seems like new customers purchase more expensive items on average compared to loyal and regular customers.
- Loyal customers appear to purchase more lower-cost items on average, which could be due to familiarity (knowing when items go on sale, using discounts, etc.)
- Regular customers are in the middle with an average of about $12.50 per item.

### Step 6 - The team wants to target different types of spenders in their marketing campaign. Create a spending flag for each user based on the average price across their orders.

In [24]:
# First creating a column containing the average prices of items users buy. 

df_ords_prods_merge['avg_product_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform('mean')

In [25]:
# Checking only specific columnns.

df_ords_prods_merge[['user_id', 'prices', 'avg_product_price']].head(15)

user_id  prices  avg_product_price
0       138     5.8           6.935811
1       138     5.8           6.935811
2       709     5.8           7.930208
3       764     5.8           4.972414
4       764     5.8           4.972414
5       777     5.8           6.935398
6       825     5.8           5.957576
7       910     5.8           6.680000
8      1052     5.8           7.162500
9      1052     5.8           7.162500
10     1379     5.8           5.813636
11     1480     5.8           5.081633
12     1480     5.8           5.081633
13     1480     5.8           5.081633
14     1494     5.8           6.803797

In [26]:
# Using conditions to create spending flags using the loc function. 

df_ords_prods_merge.loc[df_ords_prods_merge['avg_product_price'] < 10, 'spending_flag'] = 'Low spender'

In [27]:
df_ords_prods_merge.loc[df_ords_prods_merge['avg_product_price'] >= 10, 'spending_flag'] = 'High spender'

In [28]:
# Checking frequency of flags.

df_ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

Note: The total of both flags matches the shape (output 5), which means all of the data should have a label. 

In [29]:
# Checking entire dataframe.

df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  reordered  _merge    price_range_loc   busiest_day    busiest_days  \
0  ...          0    both  Mid-range product  Regular busy  Regularly busy   
1  ...          1    both  Mid-range product  Regular busy  Regularly busy   
2  ...          0    both  Mid-range product   Busiest day    Busiest days   
3  ...          0    both  Mid-range product  Regular busy    Slowest days   
4  ...          1    both  Mid-range product    Least busy    Slowest days   

  busiest_period_of_day max_order      loyalty_flag  avg_product_price  \
0           Most orders        32  Regular customer           6.935811   
1        Average orders        32  Regular customer           6.935811   
2        Average orders         5      New customer           7.930208   
3           Most orders         3      New customer           4.972414   
4        Average orders         3      New customer           4.972414   

  spending_flag  
0   Low spender  
1   Low spender  
2   Low spender  
3   Low spender  
4   Low spender  

[5 rows x 22 columns]

### Step 7 - Create an order frequency flag that marks the regularity of a user's ordering behavior according to the median in the 'days_since_prior_order' column.

In [30]:
# Frequent vs. Non-frequent customers and their ordering behavior. 
# Finding the median in 'days_since_prior_order' column for each user and grouping by user.
# Creating a column to put those results into.

df_ords_prods_merge['median_days_between_orders'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [31]:
# Checking dataframe.

df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  _merge    price_range_loc   busiest_day    busiest_days  \
0  ...    both  Mid-range product  Regular busy  Regularly busy   
1  ...    both  Mid-range product  Regular busy  Regularly busy   
2  ...    both  Mid-range product   Busiest day    Busiest days   
3  ...    both  Mid-range product  Regular busy    Slowest days   
4  ...    both  Mid-range product    Least busy    Slowest days   

  busiest_period_of_day max_order      loyalty_flag avg_product_price  \
0           Most orders        32  Regular customer          6.935811   
1        Average orders        32  Regular customer          6.935811   
2        Average orders         5      New customer          7.930208   
3           Most orders         3      New customer          4.972414   
4        Average orders         3      New customer          4.972414   

   spending_flag median_days_between_orders  
0    Low spender                        8.0  
1    Low spender                        8.0  
2    Low spender                        8.0  
3    Low spender                        9.0  
4    Low spender                        9.0  

[5 rows x 23 columns]

In [32]:
# Creating frequency flags based on the median column created and the conditions listed. 

df_ords_prods_merge.loc[df_ords_prods_merge['median_days_between_orders'] > 20, 'ordering_behavior'] = 'Non-frequent customer'

In [33]:
df_ords_prods_merge.loc[(df_ords_prods_merge['median_days_between_orders'] > 10) & (df_ords_prods_merge ['median_days_between_orders'] <= 20), 'ordering_behavior'] = 'Regular customer'

In [34]:
df_ords_prods_merge.loc[df_ords_prods_merge['median_days_between_orders'] <= 10, 'ordering_behavior'] = 'Frequent customer'

In [35]:
# Checking count of flags.

df_ords_prods_merge['ordering_behavior'].value_counts(dropna = False)

ordering_behavior
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [36]:
# Looking at the missing values in the flag column.

df_nan = df_ords_prods_merge[df_ords_prods_merge['ordering_behavior'].isnull() == True]

In [37]:
df_nan

product_id                                       product_name  \
6234909        10749                            Organic Red Bell Pepper   
12947653       21334                              Organic Peeled Garlic   
13839012       22198  4X Ultra Concentrated Natural Laundry Detergen...   
14758536       23695                           California Veggie Burger   
21673807       33401                               Goat Cheese Crumbles   

          aisle_id  department_id  prices  order_id  user_id  order_number  \
6234909         83              4     5.7    895835   159838             1   
12947653       123              4    10.2    895835   159838             1   
13839012        75             17     1.7    895835   159838             1   
14758536        42              1     4.7    895835   159838             1   
21673807        21             16    14.8    895835   159838             1   

          orders_day_of_week  order_hour_of_day  ...    price_range_loc  \
6234909                    0                 17  ...  Mid-range product   
12947653                   0                 17  ...  Mid-range product   
13839012                   0                 17  ...  Low-range product   
14758536                   0                 17  ...  Low-range product   
21673807                   0                 17  ...  Mid-range product   

          busiest_day  busiest_days busiest_period_of_day max_order  \
6234909   Busiest day  Busiest days        Average orders         1   
12947653  Busiest day  Busiest days        Average orders         1   
13839012  Busiest day  Busiest days        Average orders         1   
14758536  Busiest day  Busiest days        Average orders         1   
21673807  Busiest day  Busiest days        Average orders         1   

          loyalty_flag avg_product_price spending_flag  \
6234909   New customer              7.42   Low spender   
12947653  New customer              7.42   Low spender   
13839012  New customer              7.42   Low spender   
14758536  New customer              7.42   Low spender   
21673807  New customer              7.42   Low spender   

          median_days_between_orders ordering_behavior  
6234909                          NaN               NaN  
12947653                         NaN               NaN  
13839012                         NaN               NaN  
14758536                         NaN               NaN  
21673807                         NaN               NaN  

[5 rows x 24 columns]

In [ ]:
df_nan[['

In [38]:
df_nan.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5 entries, 6234909 to 21673807
Data columns (total 24 columns):
 #   Column                      Non-Null Count  Dtype   
---  ------                      --------------  -----   
 0   product_id                  5 non-null      int64   
 1   product_name                5 non-null      object  
 2   aisle_id                    5 non-null      int64   
 3   department_id               5 non-null      int64   
 4   prices                      5 non-null      float64 
 5   order_id                    5 non-null      int64   
 6   user_id                     5 non-null      int64   
 7   order_number                5 non-null      int64   
 8   orders_day_of_week          5 non-null      int64   
 9   order_hour_of_day           5 non-null      int64   
 10  days_since_prior_order      0 non-null      float64 
 11  add_to_cart_order           5 non-null      int64   
 12  reordered                   5 non-null      int64   
 13  _merge          

In [41]:
with pd.option_context('display.max_rows', None):
    print(df_nan)

          product_id                                       product_name  \
6234909        10749                            Organic Red Bell Pepper   
12947653       21334                              Organic Peeled Garlic   
13839012       22198  4X Ultra Concentrated Natural Laundry Detergen...   
14758536       23695                           California Veggie Burger   
21673807       33401                               Goat Cheese Crumbles   

          aisle_id  department_id  prices  order_id  user_id  order_number  \
6234909         83              4     5.7    895835   159838             1   
12947653       123              4    10.2    895835   159838             1   
13839012        75             17     1.7    895835   159838             1   
14758536        42              1     4.7    895835   159838             1   
21673807        21             16    14.8    895835   159838             1   

          orders_day_of_week  order_hour_of_day  ...    price_range_loc  \
62349

In [42]:
# renaming column 

df_ords_prods_merge = df_ords_prods_merge.rename(columns={'ordering_behavior': 'ordering_behavior_flag'})

In [43]:
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...    price_range_loc   busiest_day    busiest_days busiest_period_of_day  \
0  ...  Mid-range product  Regular busy  Regularly busy           Most orders   
1  ...  Mid-range product  Regular busy  Regularly busy        Average orders   
2  ...  Mid-range product   Busiest day    Busiest days        Average orders   
3  ...  Mid-range product  Regular busy    Slowest days           Most orders   
4  ...  Mid-range product    Least busy    Slowest days        Average orders   

  max_order      loyalty_flag avg_product_price spending_flag  \
0        32  Regular customer          6.935811   Low spender   
1        32  Regular customer          6.935811   Low spender   
2         5      New customer          7.930208   Low spender   
3         3      New customer          4.972414   Low spender   
4         3      New customer          4.972414   Low spender   

   median_days_between_orders ordering_behavior_flag  
0                         8.0      Frequent customer  
1                         8.0      Frequent customer  
2                         8.0      Frequent customer  
3                         9.0      Frequent customer  
4                         9.0      Frequent customer  

[5 rows x 24 columns]

# ADD THIS TO TASK? TRYING TO FIND WHY THERE ARE 5 MISSING VALUES

In [44]:
df_nan_five = df_ords_prods_merge[df_ords_prods_merge['ordering_behavior_flag'].isnull() == True]

In [60]:
df_nan_five 

product_id                                       product_name  \
6234909        10749                            Organic Red Bell Pepper   
12947653       21334                              Organic Peeled Garlic   
13839012       22198  4X Ultra Concentrated Natural Laundry Detergen...   
14758536       23695                           California Veggie Burger   
21673807       33401                               Goat Cheese Crumbles   

          aisle_id  department_id  prices  order_id  user_id  order_number  \
6234909         83              4     5.7    895835   159838             1   
12947653       123              4    10.2    895835   159838             1   
13839012        75             17     1.7    895835   159838             1   
14758536        42              1     4.7    895835   159838             1   
21673807        21             16    14.8    895835   159838             1   

          orders_day_of_week  order_hour_of_day  ...    price_range_loc  \
6234909                    0                 17  ...  Mid-range product   
12947653                   0                 17  ...  Mid-range product   
13839012                   0                 17  ...  Low-range product   
14758536                   0                 17  ...  Low-range product   
21673807                   0                 17  ...  Mid-range product   

          busiest_day  busiest_days busiest_period_of_day max_order  \
6234909   Busiest day  Busiest days        Average orders         1   
12947653  Busiest day  Busiest days        Average orders         1   
13839012  Busiest day  Busiest days        Average orders         1   
14758536  Busiest day  Busiest days        Average orders         1   
21673807  Busiest day  Busiest days        Average orders         1   

          loyalty_flag avg_product_price spending_flag  \
6234909   New customer              7.42   Low spender   
12947653  New customer              7.42   Low spender   
13839012  New customer              7.42   Low spender   
14758536  New customer              7.42   Low spender   
21673807  New customer              7.42   Low spender   

          median_days_between_orders ordering_behavior_flag  
6234909                          NaN                    NaN  
12947653                         NaN                    NaN  
13839012                         NaN                    NaN  
14758536                         NaN                    NaN  
21673807                         NaN                    NaN  

[5 rows x 24 columns]

In [46]:
df_nan_five[['ordering_behavior_flag', 'median_days_between_orders', 'days_since_prior_order', 'user_id']]

ordering_behavior_flag  median_days_between_orders  \
6234909                     NaN                         NaN   
12947653                    NaN                         NaN   
13839012                    NaN                         NaN   
14758536                    NaN                         NaN   
21673807                    NaN                         NaN   

          days_since_prior_order  user_id  
6234909                      NaN   159838  
12947653                     NaN   159838  
13839012                     NaN   159838  
14758536                     NaN   159838  
21673807                     NaN   159838

In [47]:
df_ords_prods_merge.isnull().sum()

# Know that the 2076096 "missing" are not actually missing, but represent first orders.
# Don't know why there are 5 missing in the mdspo and obf columns.

product_id                          0
product_name                        0
aisle_id                            0
department_id                       0
prices                              0
order_id                            0
user_id                             0
order_number                        0
orders_day_of_week                  0
order_hour_of_day                   0
days_since_prior_order        2076096
add_to_cart_order                   0
reordered                           0
_merge                              0
price_range_loc                     0
busiest_day                         0
busiest_days                        0
busiest_period_of_day               0
max_order                           0
loyalty_flag                        0
avg_product_price                   0
spending_flag                       0
median_days_between_orders          5
ordering_behavior_flag              5
dtype: int64

In [48]:

df_ords_prods_merge['ordering_behavior_flag'].value_counts(dropna = False)

ordering_behavior_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

- check the number of missing with number in exercise or check excel file to see if that is the same number.
- check code - although others also had five missing.
- just note and tell tutor

In [50]:
# Checking missing in 'days_since_prior_order' column

df_nan_days_since_prior_order = df_ords_prods_merge['days_since_prior_order'].isnull()
df_nan_days_since_prior_order

0           False
1           False
2           False
3            True
4           False
            ...  
32404854    False
32404855     True
32404856    False
32404857    False
32404858    False
Name: days_since_prior_order, Length: 32404859, dtype: bool

In [53]:
df_nan_days_since = df_ords_prods_merge[df_ords_prods_merge['days_since_prior_order'].isnull() == True]

In [56]:
df_nan_days_since[['ordering_behavior_flag', 'median_days_between_orders', 'days_since_prior_order', 'user_id']]

ordering_behavior_flag  median_days_between_orders  \
3             Frequent customer                         9.0   
10             Regular customer                        20.0   
34            Frequent customer                         2.0   
37        Non-frequent customer                        25.0   
42             Regular customer                        13.0   
...                         ...                         ...   
32404829       Regular customer                        17.0   
32404834       Regular customer                        20.0   
32404839      Frequent customer                         0.0   
32404849       Regular customer                        12.0   
32404855  Non-frequent customer                        30.0   

          days_since_prior_order  user_id  
3                            NaN      764  
10                           NaN     1379  
34                           NaN     1598  
37                           NaN     1647  
42                           NaN     2833  
...                          ...      ...  
32404829                     NaN   159487  
32404834                     NaN   163632  
32404839                     NaN   166213  
32404849                     NaN   184081  
32404855                     NaN   200377  

[2076096 rows x 4 columns]

In [57]:
df_nan_days_since['days_since_prior_order']

3          NaN
10         NaN
34         NaN
37         NaN
42         NaN
            ..
32404829   NaN
32404834   NaN
32404839   NaN
32404849   NaN
32404855   NaN
Name: days_since_prior_order, Length: 2076096, dtype: float64

In [59]:
with pd.option_context('display.max_rows', None):
    print(df_nan_days_since['days_since_prior_order'])

IOPub data rate exceeded.
The Jupyter server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--ServerApp.iopub_data_rate_limit`.

Current values:
ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
ServerApp.rate_limit_window=3.0 (secs)

